Mathlib/Order/Partition/Finpartition.lean

```diff
@@ -0,0 +1,618 @@
+            · rwa [hz2.2 _ (hQ hi), ← hz1.2 _ (hQ hi)]
+            · rwa [hz1.2 _ (hR hi), ← hz2.2 _ (hR hi)]))
+    (by
+      refine' (Finset.sup_le fun t ht ↦ _).antisymm fun a ha ↦ _
+      · rw [mem_image] at ht
+        obtain ⟨A, _, rfl⟩ := ht
+        exact s.filter_subset _
+      · rw [mem_sup]
+        refine'
+          ⟨s.filter fun i ↦ ∀ t, t ∈ F → ((t ∈ F.filter fun u ↦ a ∈ u) ↔ i ∈ t),
+            mem_image_of_mem _ (mem_powerset.2 <| filter_subset _ _),
+            mem_filter.2 ⟨ha, fun t ht ↦ _⟩⟩
+        rw [mem_filter]
+        exact and_iff_right ht)
+#align finpartition.atomise Finpartition.atomise
+
+variable {F : Finset (Finset α)}
+
+-- porting note:
+/- ./././Mathport/Syntax/Translate/Basic.lean:632:2: warning: expanding binder collection
+   (Q «expr ⊆ » F) -/
+theorem mem_atomise :
+    t ∈ (atomise s F).parts ↔
+      t.Nonempty ∧ ∃ (Q : _)(_ : Q ⊆ F), (s.filter fun i ↦ ∀ u ∈ F, u ∈ Q ↔ i ∈ u) = t := by
+  simp only [atomise, ofErase, bot_eq_empty, mem_erase, mem_image, nonempty_iff_ne_empty,
+    mem_singleton, and_comm, mem_powerset, exists_prop]
+#align finpartition.mem_atomise Finpartition.mem_atomise
+
+theorem atomise_empty (hs : s.Nonempty) : (atomise s ∅).parts = {s} := by
+  simp only [atomise, powerset_empty, image_singleton, not_mem_empty, IsEmpty.forall_iff,
+    imp_true_iff, filter_True]
+  exact erase_eq_of_not_mem (not_mem_singleton.2 hs.ne_empty.symm)
+#align finpartition.atomise_empty Finpartition.atomise_empty
+
+theorem card_atomise_le : (atomise s F).parts.card ≤ 2 ^ F.card :=
+  (card_le_of_subset <| erase_subset _ _).trans <| Finset.card_image_le.trans (card_powerset _).le
+#align finpartition.card_atomise_le Finpartition.card_atomise_le
+
+theorem bunionᵢ_filter_atomise (ht : t ∈ F) (hts : t ⊆ s) :
+    ((atomise s F).parts.filter fun u ↦ u ⊆ t ∧ u.Nonempty).bunionᵢ id = t := by
+  ext a
+  refine' mem_bunionᵢ.trans ⟨fun ⟨u, hu, ha⟩ ↦ (mem_filter.1 hu).2.1 ha, fun ha ↦ _⟩
+  obtain ⟨u, hu, hau⟩ := (atomise s F).exists_mem (hts ha)
+  refine' ⟨u, mem_filter.2 ⟨hu, fun b hb ↦ _, _, hau⟩, hau⟩
+  obtain ⟨Q, _hQ, rfl⟩ := (mem_atomise.1 hu).2
+  rw [mem_filter] at hau hb
+  rwa [← hb.2 _ ht, hau.2 _ ht]
+#align finpartition.bUnion_filter_atomise Finpartition.bunionᵢ_filter_atomise
+
+theorem card_filter_atomise_le_two_pow (ht : t ∈ F) :
+    ((atomise s F).parts.filter fun u ↦ u ⊆ t ∧ u.Nonempty).card ≤ 2 ^ (F.card - 1) := by
+  suffices h :
+    ((atomise s F).parts.filter fun u ↦ u ⊆ t ∧ u.Nonempty) ⊆
+      (F.erase t).powerset.image fun P ↦ s.filter fun i ↦ ∀ x ∈ F, x ∈ insert t P ↔ i ∈ x
+  · refine' (card_le_of_subset h).trans (card_image_le.trans _)
+    rw [card_powerset, card_erase_of_mem ht]
+  rw [subset_iff]
+  simp_rw [mem_image]
+  simp_rw [mem_powerset]
+  simp_rw [mem_filter]
+  simp_rw [and_imp]
+  simp_rw [Finset.Nonempty]
+  simp_rw [exists_imp]
+  simp_rw [mem_atomise]
+  simp_rw [and_imp]
+  simp_rw [Finset.Nonempty]
+  simp_rw [exists_imp]
```

Comment:
```suggestion
  simp_rw [mem_image, mem_powerset, mem_filter, and_imp, Finset.Nonempty,
    exists_imp, mem_atomise, and_imp,Finset.Nonempty, exists_imp]
```

Reply:
Hmm, that loses the ability to look at the individual steps easily (wasn’t that different in lean3? Or only with `rw`), but it works.